community/index/src/main/java/org/neo4j/index/internal/gbptree/Meta.java

```diff
@@ -0,0 +1,243 @@
+     *
+     * @param cursor {@link PageCursor} to read meta information from.
+     * @param layout {@link Layout} instance that will get the opportunity to read layout-specific data from the meta page.
+     * {@code layout} is allowed to be {@code null} where it won't be told to read layout-specific data from the meta page.
+     * @return {@link Meta} instance with all meta information.
+     * @throws IOException on {@link PageCursor} I/O error.
+     */
+    static Meta read( PageCursor cursor, Layout<?,?> layout ) throws IOException
+    {
+        int format;
+        int pageSize;
+        long layoutIdentifier;
+        int layoutMajorVersion;
+        int layoutMinorVersion;
+        try
+        {
+            do
+            {
+                format = cursor.getInt();
+                pageSize = cursor.getInt();
+                layoutIdentifier = cursor.getLong();
+                layoutMajorVersion = cursor.getInt();
+                layoutMinorVersion = cursor.getInt();
+                if ( layout != null )
+                {
+                    layout.readMetaData( cursor );
+                }
+            }
+            while ( cursor.shouldRetry() );
+            checkOutOfBounds( cursor );
+            cursor.checkAndClearCursorException();
+        }
+        catch ( CursorException e )
+        {
+            throw new MetadataMismatchException( e,
+                    "Tried to open, but caught an error while reading meta data. " +
+                    "File is expected to be corrupt, try to rebuild." );
+        }
+
+        return parseMeta( format, pageSize, layoutIdentifier, layoutMajorVersion, layoutMinorVersion );
+    }
+
+    void verify( Layout<?,?> layout )
+    {
+        if ( unusedVersionSlot3 != Meta.UNUSED_VERSION )
+        {
+            throw new MetadataMismatchException( "Unexpected version " + unusedVersionSlot3 + " for unused version slot 3" );
+        }
+        if ( unusedVersionSlot4 != Meta.UNUSED_VERSION )
+        {
+            throw new MetadataMismatchException( "Unexpected version " + unusedVersionSlot4 + " for unused version slot 4" );
+        }
+
+        if ( !layout.compatibleWith( layoutIdentifier, layoutMajorVersion, layoutMinorVersion ) )
+        {
+            throw new MetadataMismatchException(
+                    "Tried to open using layout not compatible with " +
+                    "what the index was created with. Created with: layoutIdentifier=%d,majorVersion=%d,minorVersion=%d. " +
+                    "Opened with layoutIdentifier=%d,majorVersion=%d,minorVersion=%d",
+                    layoutIdentifier, layoutMajorVersion, layoutMinorVersion,
+                    layout.identifier(), layout.majorVersion(), layout.minorVersion() );
+        }
+
+        Factory formatByLayout = TreeNodeSelector.selectByLayout( layout );
+        if ( formatByLayout.formatIdentifier() != formatIdentifier ||
+             formatByLayout.formatVersion() != formatVersion )
```

Comment:
True, however this isn't a problem that we have right now. Imho right now what's important is that we store the identifier and version and then we can adapt the checking around it anyway we like when it gets to that. Guessing how it will work in the future will most likely end up in something that needs to be changed anyway.